File: 03/01_busca/codes/busca.c
#include <stdio.h>

#define N 1000000

void preenche_vetor(long v[], long n){
  long i;
  long j;
  for(i=0,j=0; j<n;i+=2) v[j++] = i;
}

void imprime_vetor(long v[], long n){
  long i;
  for(i=0; i<n;i++) printf("v[%ld] = %ld\n",i,v[i]);
}

long busca_sequencial(long v[], long n, long elem){
  long i;
  for(i=0; i<n;i++)
    if (v[i] == elem){
      printf("%ld comparacoes\n",i);
      return i;
    }
  printf("%ld comparacoes\n",i);
  return(-1);
}

long busca_binaria(long v[], long n, long elem){
  long l,r,m,comp=0;
  l = 0;
  r = n-1;

  while(l <= r){
    m = (l+r)/2;
    if(v[m] == elem)
    {
      printf("%ld comparacoes\n",comp);
      return(m);
    }
    else if(v[m] < elem)
      l = m + 1;
    else if(v[m] > elem)
      r = m - 1;
    comp++;
  }
  printf("%ld comparacoes\n",comp);
  return(-1);
}

long busca_binaria_recur(long v[], long n, long elem, long l, long r){
  long m = (l+r)/2;

  if(l > r)
    return(-1);

  if(v[m] == elem)
    return(m);
  else if(v[m] < elem)
    return(busca_binaria_recur(v,n,elem,m+1,r));
  else if(v[m] > elem)
    return(busca_binaria_recur(v,n,elem,l,m-1));
}

int main(){
  long v[N],x,r;

  preenche_vetor(v,N);
  //imprime_vetor(v,N);

  scanf("%ld",&x);

  r = busca_sequencial(v,N,x);
  //r = busca_binaria(v,N,x);
  //r = busca_binaria_recur(v,N,x,0,N-1);
  if(r != -1)
    printf("O elemento %ld esta na posicao %ld do vetor.\n",x,r);
  else
    printf("O elemento %ld nao pertence ao vetor.\n",x);
}

File: 02/codes/busca/busca.go
// Busca sequencial, busca binária (iterativa) e busca binária (recursiva)
// em um vetor ordenado de 1.000.000 de elementos, para comparar o número
// de comparações realizado por cada estratégia.

package main

import "fmt"

const n = 1_000_000

func preencheVetor(n int) []int {
	v := make([]int, n)
	for i, j := 0, 0; j < n; i += 2 {
		v[j] = i
		j++
	}
	return v
}

// buscaSequencial percorre o vetor do início ao fim.
// Custo: O(n) no pior caso.
func buscaSequencial(v []int, elem int) int {
	i := 0
	for ; i < len(v); i++ {
		if v[i] == elem {
			fmt.Printf("%d comparações (sequencial)\n", i+1)
			return i
		}
	}
	fmt.Printf("%d comparações (sequencial)\n", i)
	return -1
}

// buscaBinaria divide o intervalo de busca pela metade a cada passo.
// Custo: O(log n) no pior caso.
func buscaBinaria(v []int, elem int) int {
	l, r := 0, len(v)-1
	comp := 0

	for l <= r {
		m := (l + r) / 2
		comp++
		if v[m] == elem {
			fmt.Printf("%d comparações (binária iterativa)\n", comp)
			return m
		} else if v[m] < elem {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	fmt.Printf("%d comparações (binária iterativa)\n", comp)
	return -1
}

// buscaBinariaRecursiva é a mesma estratégia da busca binária, escrita
// recursivamente em vez de com um laço.
func buscaBinariaRecursiva(v []int, elem, l, r int) int {
	if l > r {
		return -1
	}

	m := (l + r) / 2
	if v[m] == elem {
		return m
	} else if v[m] < elem {
		return buscaBinariaRecursiva(v, elem, m+1, r)
	}
	return buscaBinariaRecursiva(v, elem, l, m-1)
}

func imprimeResultado(nome string, elem, r int) {
	if r != -1 {
		fmt.Printf("[%s] O elemento %d está na posição %d do vetor.\n", nome, elem, r)
	} else {
		fmt.Printf("[%s] O elemento %d não pertence ao vetor.\n", nome, elem)
	}
}

func main() {
	v := preencheVetor(n)

	var x int
	fmt.Scan(&x)

	imprimeResultado("sequencial", x, buscaSequencial(v, x))
	imprimeResultado("binária iterativa", x, buscaBinaria(v, x))
	imprimeResultado("binária recursiva", x, buscaBinariaRecursiva(v, x, 0, n-1))
}
